ably/realtime_channel.go

```diff
@@ -110,61 +110,67 @@ func (ch *Channels) All() []*RealtimeChannel {
 func (ch *Channels) Release(name string) error {
 	ch.mtx.Lock()
 	defer ch.mtx.Unlock()
-	if c, ok := ch.chans[name]; ok {
-		delete(ch.chans, name)
-		return c.Close()
+	c, ok := ch.chans[name]
+	if !ok {
+		return nil
 	}
+	err := wait(c.Detach())
+	if err != nil {
+		return err
+	}
+	delete(ch.chans, name)
 	return nil
 }
 
-func (ch *Channels) broadcastConnStateChange(state State) {
+func (ch *Channels) broadcastConnStateChange(change ConnectionStateChange) {
 	ch.mtx.Lock()
 	defer ch.mtx.Unlock()
 	for _, c := range ch.chans {
-		c.onConnState(state)
+		c.onConnStateChange(change)
 	}
 }
 
 // RealtimeChannel represents a single named message channel.
 type RealtimeChannel struct {
+	mtx sync.Mutex
+
+	ChannelEventEmitter
 	Name     string            // name used to create the channel
 	Presence *RealtimePresence //
 
+	state           ChannelState
+	errorReason     *ErrorInfo
+	internalEmitter ChannelEventEmitter
```

Comment:
Isn't this redundant? We are already embedding `ChannelEventEmitter`. Which we only use once  in `RealtimeChannel.lockSetState` in which we emit the same event twice. I think we can just use one emitter per `RealtimeChannel` to simplify things.

Reply:
The thing is we want separate emitters for users and for the internals of the library. We don't want them to affect each other in any way. Specifically, we don't want [`off` called with no listener](https://docs.ably.io/client-lib-development-guide/features/#RTE5) (which deregisters all listeners) to affect internal listeners.

(I'm now realizing we're missing the `Off` and `OfAll` methods with no listeners. Will add in a new PR.)

Reply:
This behavior is specified in [RTC10](https://docs.ably.io/client-lib-development-guide/features/#RTC10): "The client library should never register any listeners for internal use with the public EventEmitter interfaces"